src/shims/sync.rs

```diff
@@ -0,0 +1,438 @@
+        let locked_count = this.read_scalar(locked_count_place.into())?.to_u32()?;
+
+        if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")? {
+            if locked_count == 1 {
+                this.write_scalar(Scalar::from_u32(0), locked_count_place.into())?;
+                Ok(0)
+            } else {
+                throw_ub_format!("Attempted to unlock a PTHREAD_MUTEX_NORMAL mutex that was not locked");
+            }
+        } else if kind == this.eval_libc("PTHREAD_MUTEX_ERRORCHECK")? {
+            if locked_count == 1 {
+                this.write_scalar(Scalar::from_u32(0), locked_count_place.into())?;
+                Ok(0)
+            } else {
+                this.eval_libc_i32("EPERM")
+            }
+        } else if kind == this.eval_libc("PTHREAD_MUTEX_RECURSIVE")? {
+            if locked_count > 0 {
+                this.write_scalar(Scalar::from_u32(locked_count - 1), locked_count_place.into())?;
+                Ok(0)
+            } else {
+                this.eval_libc_i32("EPERM")
+            }
+        } else {
+            this.eval_libc_i32("EINVAL")
+        }
+    }
+
+    fn pthread_mutex_destroy(&mut self, mutex_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, mutex_op, 16)?;
+
+        let mutex = this.read_scalar(mutex_op)?.not_undef()?;
+        if this.is_null(mutex)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+        let mutex_place = this.deref_operand(mutex_op)?;
+
+        let u32_layout = this.layout_of(this.tcx.types.u32)?;
+        let locked_count_place = mutex_place.offset(Size::from_bytes(4), MemPlaceMeta::None, u32_layout, this)?;
+        if this.read_scalar(locked_count_place.into())?.to_u32()? != 0 {
+            return this.eval_libc_i32("EBUSY");
+        }
+
+        let i32_layout = this.layout_of(this.tcx.types.i32)?;
+        let kind_place = mutex_place.offset(Size::from_bytes(12), MemPlaceMeta::None, i32_layout, this)?;
+        this.write_scalar(ScalarMaybeUndef::Undef, kind_place.into())?;
+        this.write_scalar(ScalarMaybeUndef::Undef, locked_count_place.into())?;
+
+        Ok(0)
+    }
+
+    // pthread_rwlock_t is between 32 and 56 bytes, depending on the platform
+    // memory layout:
+    // bytes 0-3: reserved for signature on macOS
+    // bytes 4-7: reader count, as a u32
+    // bytes 8-11: writer count, as a u32
+
+    fn pthread_rwlock_rdlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, rwlock_op, 12)?;
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+        let rwlock_place = this.deref_operand(rwlock_op)?;
+
+        let u32_layout = this.layout_of(this.tcx.types.u32)?;
```

Comment:
This looks like it is time to do what others have proposed before -- to add a thing somewhere in our machine state (`Evaluator` or `MemoryExtra`) where we store the layout of some of these frequently used types, just so that we can readily access them everywhere.